algorithms/indexing/lattice_search_strategies.py

```diff
@@ -0,0 +1,611 @@
+            cand_rlp = matrix.col(self.spots[cand["spot_id"]]["rlp"])
+            cand_vec = cand["rlp_datum"]
+
+            obs_dists = [(cand_rlp - rlp).length() for rlp in obs_relps]
+            exp_dists = [(vec - cand_vec).length() for vec in exp_relps]
+
+            residual_dist = [abs(a - b) for (a, b) in zip(obs_dists, exp_dists)]
+
+            # If any of the distance differences is larger than the sum in quadrature
+            # of the tolerated d* bands then reject the candidate
+            sq_candidate_band = self.spots[cand["spot_id"]]["dstar_band2"]
+            bad_candidate = False
+            for r_dist, spot_id in zip(residual_dist, existing_ids):
+                sq_relp_band = self.spots[spot_id]["dstar_band2"]
+                if r_dist > sqrt(sq_relp_band + sq_candidate_band):
+                    bad_candidate = True
+                    break
+            if bad_candidate:
+                continue
+
+            # Calculate co-planarity of the relps, including the origin
+            points = flex.vec3_double(exp_relps + [cand_vec, (0.0, 0.0, 0.0)])
+            plane = least_squares_plane(points)
+            plane_score = flex.sum_sq(
+                points.dot(plane.normal) - plane.distance_to_origin
+            )
+
+            # Reject if the group of relps are too far from lying in a single plane.
+            # This cut-off was determined by trial and error using simulated images.
+            if plane_score > 6e-7:
+                continue
+
+            # Construct a graph including the accepted candidate node
+            g = graph.factory_add_vertex(
+                {
+                    "spot_id": cand["spot_id"],
+                    "miller_index": cand["miller_index"],
+                    "rlp_datum": cand["rlp_datum"],
+                },
+                weights_to_other=residual_dist,
+            )
+
+            result.append(g)
+
+        return result
+
+    @staticmethod
+    def _fit_U_from_superposed_points(reference, other):
+
+        # Add the origin to both sets of points
+        origin = flex.vec3_double(1)
+        reference.extend(origin)
+        other.extend(origin)
+
+        # Find U matrix that takes ideal relps to the reference
+        fit = superpose.least_squares_fit(reference, other)
+        return fit.r
+
+    def _fit_crystal_model(self, graph):
+
+        vertices = graph.vertices
+
+        # Reciprocal lattice points of the observations
+        sel = flex.size_t([e["spot_id"] for e in vertices])
+        reference = self.spots["rlp"].select(sel)
+
+        # Ideal relps from the known cell
+        other = flex.vec3_double([e["rlp_datum"] for e in vertices])
+
+        U = self._fit_U_from_superposed_points(reference, other)
+        UB = U * self.Bmat
+
+        if self._params.bootstrap_crystal:
+
+            # Attempt to index the low resolution spots
+            from dials_algorithms_indexing_ext import AssignIndices
```

Comment:
Ideally I'd prefer not to import directly from the extension here, but I see the current implementation exposed in https://github.com/dials/dials/blob/master/algorithms/indexing/assign_indices.py expects reflections/experiments as input, which you don't currently have at this point...

Reply:
Yes, that's the issue. I'm not yet sure how useful the non-default option to trigger this import is. It seemed to occasionally help, but not enough to make it default